US585006: A new API to check status of the batch

raghavendran-gopalakrishnan opened the pull request:
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=585006

raghavendran-gopalakrishnan committed:
US585006: A new API to check status of the batch

https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=585006

raghavendran-gopalakrishnan committed:
Addressed review comments

raghavendran-gopalakrishnan committed:
Changed to return Enum as status from Dao

kusumaghoshdastidar committed:
Un related change

kusumaghoshdastidar committed:
Refactor batch not found state

kusumaghoshdastidar committed:
Update test verification

kusumaghoshdastidar committed:
Extract batch id from in-progress batch folder name

kusumaghoshdastidar committed:
Fix in progress batch folder name check

kusumaghoshdastidar committed:
Add test dependency

kusumaghoshdastidar committed:
Add license header

raghavendran-gopalakrishnan committed:
Added functionality to check if batch is stale

raghavendran-gopalakrishnan committed:
Updated tests

raghavendran-gopalakrishnan committed:
Merge branch 'develop' into US585006

raghavendran-gopalakrishnan committed:
Removed the test files

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=CAFDataProcessing&REPO_NAME=staging-service&BRANCH_NAME=US585006) to go ahead and create the build...

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFDataProcessing/job/CAFDataProcessing~staging-service~US585006~CI

dermot-hardy commented on staging-service-shared/pom.xml:
```diff
@@ -33,6 +33,17 @@
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>${junit.jupiter.version}</version>
```
You added this to dependency management so you shouldn't be listing the version again here.

dermot-hardy commented on staging-service-shared/pom.xml:
```diff
@@ -33,6 +33,17 @@
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>${junit.jupiter.version}</version>
```
You didn't add this to dependency management but why not?  You correctly added the other one.

dermot-hardy commented on staging-service-shared/pom.xml:
```diff
@@ -33,6 +33,17 @@
+            <scope>test</scope>
+        </dependency>
     </dependencies>
-
```
Unnecessary whitespace change.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -341,4 +356,11 @@ private boolean shouldDelete(final String filename)
+
+    private boolean isInProgressBatchAbandoned(final Path path, final BatchId batchId)
+    {
+        return Arrays.stream(path.toFile().list())
```
Don't use any method without reading the JavaDoc for it.  In this case it literally talks about a new method that should be used instead which uses less resources and is more responsive.

![image](https://user-images.githubusercontent.com/20627297/204958150-18cf0385-fa2e-4f78-ae5d-f0ce4068fafa.png)

It's also worth considering [`Files.list()`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#list-java.nio.file.Path-) since I see it is used elsewhere in this file.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +283,20 @@ public void cleanUpStaleInprogressBatches()
+    {
+        if (batchPathProvider.getPathForBatch(tenantId, batchId).toFile().exists()) {
+            return BatchStatus.COMPLETED;
+        } else if (skipBatchFileCleanup && isInProgressBatchAbandoned(batchPathProvider.getTenantInprogressDirectory(tenantId), batchId)) {
```
Why would the `skipBatchFileCleanup` flag be relevant here?

raghavgopal replied:
Because only if this flag is set to true, the cleanup batch will not be running. By default, the flag is false, the batch runs and there will not be any abandoned batches.

dermot-hardy replied:
You're assuming there isn't a second Staging Service that has it configured on.  It only takes one Staging Service instance to clean up for all.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +110,14 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
         return getPathForTenant(tenantId).resolve(INPROGRESS_FOLDER);
     }
 
+    public boolean isBatchInProgress(final TenantId tenantId, final BatchId batchId)
```
This class is called `BatchPathProvider`.  This method doesn't seem to fit with the purpose of the class.  Note that there are no other methods in here searching or querying the filesystem.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +110,14 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+    {
+        final Path pathForBatches = getTenantInprogressDirectory(tenantId);
+        return Arrays
+            .stream(pathForBatches.toFile().list())
```
Same comment on using `File.list()` again.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchNameProvider.java:
```diff
@@ -21,6 +21,7 @@
 import java.time.ZoneOffset;
 import java.time.format.DateTimeFormatter;
 import java.time.format.DateTimeParseException;
+import java.time.temporal.ChronoUnit;
```
Apparently this isn't used:

![image](https://user-images.githubusercontent.com/20627297/204960677-d405b27d-0a2f-472c-a031-ce045c2dbb1e.png)

If whatever IDE you're using doesn't point out obvious issues like this one then you need to change IDEs.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchNameProvider.java:
```diff
@@ -58,6 +59,17 @@ public static long getFileCreationTime(final String fileName)
+        final String noCreationTimeFileName = fileName.substring(23);
+
+        final int endOfThreadId = noCreationTimeFileName.indexOf('-');
+        final String noThreadIdFileName = noCreationTimeFileName.substring(endOfThreadId + 1);
```
These first two `.substring()`s are not necessary as `.indexOf()` has an overload that takes a starting index.

dermot-hardy replied:
Overlooked comment from last review:
These first two `.substring()`s are not necessary as `.indexOf()` has an overload that takes a starting index.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -341,4 +356,11 @@ private boolean shouldDelete(final String filename)
+    private boolean isInProgressBatchAbandoned(final Path path, final BatchId batchId)
+    {
+        return Arrays.stream(path.toFile().list())
+                    .filter(this::shouldDelete)
```
You're declaring the batch abandoned on the basis of this (poorly named) `shouldDelete()` method.  The actual deletion code checks much more than just this.  You could easily be declaring a batch to be abandoned here which is not.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +283,20 @@ public void cleanUpStaleInprogressBatches()
+        if (batchPathProvider.getPathForBatch(tenantId, batchId).toFile().exists()) {
+            return BatchStatus.COMPLETED;
+        } else if (skipBatchFileCleanup && isInProgressBatchAbandoned(batchPathProvider.getTenantInprogressDirectory(tenantId), batchId)) {
+            return BatchStatus.ABANDONED;
```
If you find any abandoned batch you are declaring the batch abandoned.  But surely this batch upload could have been re-started and be actively in-progress.  The fact that there was a previous abandoned attempt doesn't mean that we shouldn't report that there is a batch being actively progressed, if there is one of course.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/StagingController.java:
```diff
@@ -132,6 +132,39 @@ public ResponseEntity<Void> deleteBatch(
+            if (statusResponse == BatchDao.BatchStatus.COMPLETED) {
+                status.setMessage("Staging batch '" + batchId + "' is completed.");
+            } else if (statusResponse == BatchDao.BatchStatus.INPROGRESS) {
+                status.setMessage("Staging batch '" + batchId + "' is in progress.");
```
Completed and abandoned are basic states that don't require much testimony, but just saying that a batch is in-progress really doesn't seem sufficient.  What metrics can we provide to demonstrate that it is still in-progress?  If the service id of the in-progress batch is the same as the service id of this request then presumably we should know everything about the progress: the number of bytes received, the time the last byte was received, the rate that the bytes are being received, etc.  If it is a different service instance that is receiving the batch then naturally we only know what we can observe from the filesystem, but that should still tell us the last modification date of the last subbatch file written and so on.  We could even delay returning for a second or two to observe that there are indeed bytes being received.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/StagingController.java:
```diff
@@ -132,6 +132,39 @@ public ResponseEntity<Void> deleteBatch(
+            } else if (statusResponse == BatchDao.BatchStatus.INPROGRESS) {
+                status.setMessage("Staging batch '" + batchId + "' is in progress.");
+            } else if (statusResponse == BatchDao.BatchStatus.ABANDONED) {
+                status.setMessage("Staging batch '" + batchId + "' is abandoned.");
```
We could probably get away with this for abandoned batches but we could do much better.  How many abandoned batches do we have for this batch id?  For each of them, what was their start time (which we can easily get from the filename)?  Could we also report on the last modification time to know the time of abandonment?  I would also like if we reported on the abandoned batches even if there is a current batch in progress.

raghavendran-gopalakrishnan committed:
Added more metrics to batch status

dermot-hardy committed:
Merge branch 'develop' into US585006

dermot-hardy commented on staging-service-shared/pom.xml:
```diff
@@ -33,6 +33,19 @@
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>commons-fileupload</groupId>
+            <artifactId>commons-fileupload</artifactId>
+        </dependency>
```
Firstly compile-scope dependencies go above test-scope, but more importantly adding this dependency to the `staging-service-shared` library should have been setting off alarm bells.  This is a "shared" library.  Introducing the dependency here introduces it for all clients.  This seems unlikely to be correct.

raghavgopal replied:
Done.

dermot-hardy commented on staging-service/src/main/resources/application.yaml:
```diff
@@ -54,6 +54,8 @@ spring:
     multipart:
       enabled: false
+  jackson:
+    default-property-inclusion: non_null
```
Is this necessary?  I'd be slightly worried that this might change the behavior of existing operations.

raghavgopal replied:
Added this, to keep only not-null fields in the JSON response.

raghavgopal replied:
Existing operations will not be impacted with this change.

dermot-hardy commented on staging-service-contract/src/main/resources/com/github/cafdataprocessing/services/staging/contract/swagger.yaml:
```diff
@@ -163,4 +196,47 @@ components:
+      description: The Response object with status of the batch
+      type: object
+      properties:
+        batchID:
```
It is `batchId` in other places in the contract.

```suggestion
        batchId:
```

raghavgopal replied:
Done.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class BatchProgressTracker
```
Classes should be final unless you are intending for you to be used as a base class in an inheritance hierarchy.

raghavgopal replied:
Done.

dermot-hardy replied:
Doesn't appear to have been done for all new classes.

raghavgopal replied:
All new classes are marked final.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+        return inProgressTrackerMap;
+    }
+
+    private static final ConcurrentHashMap<String, Tracker> inProgressTrackerMap = new ConcurrentHashMap<>();
```
Put variables first in the class, before methods.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+        };
+    }
+}
+
```
Drop the extra line.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+        fileUpload.setProgressListener(createProgressListener());
+    }
+
+    private static ProgressListener createProgressListener()
```
All of this logic needs to be moved into the Staging Service.  The only reason to have it in `staging-service-shared` would be if we also wanted it to be used by the Ingestion Worker, but of course we don't want or need that.

dermot-hardy replied:
**Resolved in this commit:**  
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+
+    private static ProgressListener createProgressListener()
+    {
+        Tracker tracker = new Tracker();
```
Variables should be `final` unless there's a good reason for them not to be.  If you think you have a good reason for a mutable variable then sit and think about whether there is an alternate way the code could be written to avoid having the mutable variable.  If you can re-write it to not use mutable variables you'll normally find you have higher quality code.  If you're not in the practice of declaring variables final then you miss the opportunity to re-think about mutable variables.

dermot-hardy replied:
**Resolved in this commit:**
[9d1eb9b2f9a3aa26f5ac6c8bd742be5fb3332b66] Code review: Add missing final keywords

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+
+    private static final ConcurrentHashMap<String, Tracker> inProgressTrackerMap = new ConcurrentHashMap<>();
+
+    public static void updateTracker(ServletFileUpload fileUpload)
```
Parameters should be `final`.

raghavgopal replied:
done

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+        {
+            private long megaBytes = -1;
+            @Override
+            public void update(long totalBytesRead, long contentLength, int item) {
+                long totalBytesInMBytes = totalBytesRead / 1000000;
```
```suggestion
            public void update(final long totalBytesRead, final long contentLength, final int item)
            {
                final long totalBytesInMBytes = totalBytesRead / 1000000;
```

raghavgopal replied:
done

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/Tracker.java:
```diff
@@ -0,0 +1,77 @@
+
+public class Tracker
+{
+    private Instant lastModifiedTime;
+    private Long numberOfBytesReceived = null;
+    private Long fileUploadRate = null;
```
Use constructors for initialization.

```suggestion
    private Instant lastModifiedTime;
    private Long numberOfBytesReceived;
    private Long fileUploadRate;

    public Tracker()
    {
        this.lastModifiedTime = null;
        this.numberOfBytesReceived = null;
        this.fileUploadRate = null;
    }
```

It's more verbose but it's clearer.

raghavgopal replied:
Done

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/Tracker.java:
```diff
@@ -0,0 +1,77 @@
+                '}';
+    }
+
+    public void setFileUploadRate(long fileUploadRate)
```
Put this beside the get like the rest.  Also, the parameters of all the mutators should be `final`.

raghavgopal replied:
Done

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+
+    public static void updateTracker(ServletFileUpload fileUpload)
+    {
+        fileUpload.setProgressListener(createProgressListener());
+    }
+
+    private static ProgressListener createProgressListener()
+    {
+        Tracker tracker = new Tracker();
+        tracker.setUploadStartTime(Instant.now());
+        return new ProgressListener()
+        {
+            private long megaBytes = -1;
```
A proper class would be clearer in this case (and would also make debugging easier because the class will have a proper name):

```suggestion
        fileUpload.setProgressListener(new ProgressListenerImpl());
    }

    private static class ProgressListenerImpl implements ProgressListener
    {
        private final Tracker tracker;
        private long megaBytes;

        public ProgressListenerImpl()
        {
            final Tracker newTracker = new Tracker();
            newTracker.setUploadStartTime(Instant.now());

            this.tracker = newTracker;
            this.megaBytes = -1;
        }

```

raghavgopal replied:
Done

dermot-hardy replied:
Incidentally I moved this nested class to a top-level class and renamed it to `BatchProgressListener` in commit 590f331c0db34ae520d5a2d8a635c62a07f8bd09.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+            private long megaBytes = -1;
+            @Override
+            public void update(long totalBytesRead, long contentLength, int item) {
+                long totalBytesInMBytes = totalBytesRead / 1000000;
```
Use underscores when you have big numbers to make them easier to read:

```suggestion
                long totalBytesInMBytes = totalBytesRead / 1_000_000;
```

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+            private long megaBytes = -1;
+            @Override
+            public void update(long totalBytesRead, long contentLength, int item) {
+                long totalBytesInMBytes = totalBytesRead / 1000000;
```
Shouldn't we use 2^20 rather than 10^6?

raghavgopal replied:
Yes makes it more accurate!

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,72 @@
+        throw new IllegalStateException();
+    }
+
+    public static ConcurrentMap<String, Tracker> getInProgressTrackerMap()
+    {
+        return inProgressTrackerMap;
+    }
```
This should be encapsulated.  Rather than handing out the map and trusting callers to handle it correctly you should have appropriately named methods in this class, hiding the use of the map as an implementation detail.

raghavendran-gopalakrishnan replied:
Made the map variable `inProgressTrackerMap `as private final non-static, made the getter as non-static.

raghavgopal replied:
Done

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/StagingController.java:
```diff
@@ -96,11 +100,11 @@ public ResponseEntity<Void> createOrReplaceBatch(
         try {
             batchDao.saveFiles(new TenantId(X_TENANT_ID), new BatchId(batchId), fileItemIterator);
             LOGGER.debug("Staged batch: {}", batchId);
+            BatchProgressTracker.getInProgressTrackerMap().remove(Thread.currentThread().getId()+"-"+ ServiceIdentifier.getServiceId());
```
I don't understand removing from the map in this success case but not removing from the map in the failure catch blocks both above and below this code.  What currently happens if we go into those catch blocks and is it correct behavior?

raghavgopal replied:
Yes, I have moved that statement to finally block now.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/BatchDao.java:
```diff
@@ -22,10 +22,11 @@
 import javax.validation.Valid;
 import javax.validation.constraints.Min;
 import javax.validation.constraints.Size;
-import org.apache.commons.fileupload.FileItemIterator;
```
Don't move this

raghavgopal replied:
Done.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/BatchDao.java:
```diff
@@ -22,10 +22,11 @@
 import com.github.cafdataprocessing.services.staging.exceptions.StagingException;
 import java.util.List;
+import org.apache.commons.fileupload.FileItemIterator;
+import com.github.cafdataprocessing.services.staging.models.BatchStatusResponse;
```
Put this up in the correct location, with the other `com.*` imports.

raghavgopal replied:
Done.

dermot-hardy replied:
**Other import order issues resolved in:**  
[a30664129e09823c76988569e0d5f9f80bcf16a3] Code review: Import order

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -49,6 +49,13 @@
+import com.github.cafdataprocessing.services.staging.models.InProgressMetrics;
+import com.github.cafdataprocessing.services.staging.utils.BatchProgressTracker;
+import com.github.cafdataprocessing.services.staging.utils.ServiceIdentifier;
+import com.github.cafdataprocessing.services.staging.utils.Tracker;
```
Move up to correct location.

raghavgopal replied:
Done

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +289,49 @@ public void cleanUpStaleInprogressBatches()
+    @Override
+    public BatchStatusResponse getBatchStatus(final TenantId tenantId, final BatchId batchId) throws BatchNotFoundException
+    {
+        boolean batchInCompletedState = false;
```
A perfect example of what I was talking about in an earlier comment.  The mutable variable can be easily re-worked to be `final` and you will find that you will end up with better code.  See next comment.

raghavgopal replied:
Wow thanks for that!

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +289,49 @@ public void cleanUpStaleInprogressBatches()
+        responseBean.setBatchID(batchId.getValue());
+        BatchStatus batchStatus = new BatchStatus();
+        InProgress inProgress = new InProgress();
+        if (batchPathProvider.getPathForBatch(tenantId, batchId).toFile().exists()) {
+            batchStatus.setBatchComplete(true);
+            batchInCompletedState = true;
+        } else {
+            batchStatus.batchComplete(false);
+        }
```
```suggestion
        final boolean batchInCompletedState = batchPathProvider.getPathForBatch(tenantId, batchId).toFile().exists();
        batchStatus.setBatchComplete(batchInCompletedState);
```

raghavgopal replied:
Done

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +289,49 @@ public void cleanUpStaleInprogressBatches()
+        } else {
+            batchStatus.batchComplete(false);
+        }
+        List<InProgressMetrics> inProgressMetricsList =
+                new ArrayList<>();
```
```suggestion
        final List<InProgressMetrics> inProgressMetricsList = new ArrayList<>();
```

raghavgopal replied:
done.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+    {
+        Path inProgressPath = getTenantInprogressDirectory(tenantId);
+        List<String> list = new ArrayList<>();
+        try (Stream<Path> stream = Files.list(inProgressPath)) {
```
Variables declared in try-with-resources should also be `final`:

```suggestion
        try (final Stream<Path> stream = Files.list(inProgressPath)) {
```

raghavgopal replied:
Done.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+        try (Stream<Path> stream = Files.list(inProgressPath)) {
+            stream.filter(batch -> BatchNameProvider.getBatchId(batch.getFileName().toString()).equals(batchId.getValue()))
+                    .forEach(path -> list.add(BatchNameProvider.extractThreadIDAndServiceID(path.getFileName().toString())));
+        } catch (IOException e) {
```
And also exception catch blocks:

```suggestion
        } catch (final IOException e) {
```

raghavgopal replied:
Done

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+        List<String> list = new ArrayList<>();
+        try (Stream<Path> stream = Files.list(inProgressPath)) {
+            stream.filter(batch -> BatchNameProvider.getBatchId(batch.getFileName().toString()).equals(batchId.getValue()))
+                    .forEach(path -> list.add(BatchNameProvider.extractThreadIDAndServiceID(path.getFileName().toString())));
```
Prefer using collectors over `.forEach()`.  Using `.forEach()` requires that you use side-effects.  Also if you use a collector you can return directly from here.

raghavgopal replied:
Done.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+            stream.filter(batch -> BatchNameProvider.getBatchId(batch.getFileName().toString()).equals(batchId.getValue()))
+                    .forEach(path -> list.add(BatchNameProvider.extractThreadIDAndServiceID(path.getFileName().toString())));
+        } catch (IOException e) {
+            LOGGER.info("The folder might have been moved or deleted.");
```
I don't understand this.  The in-progress folder might have been moved or deleted?  If that is the case then it sounds like an ERROR rather than an INFO.  Also, include the exception in the log if this is indeed an unexpected circumstance.  Have you ever seen this happening?

```suggestion
            LOGGER.error("The folder might have been moved or deleted.", e);
```

I presume you're intentionally returning an empty list in this circumstance?

raghavgopal replied:
Yes, if at that exact moment when the batch completes, the in-progress folder moves to completed folder, and let's say we're checking the file system for the status, we might encounter IOException, so we are ignoring it as it's a natural exception bound to happen. And yes, returning the empty list in this circumstance.

raghavgopal replied:
However, will change them to `error` level.

dermot-hardy replied:
I'm still concerned about this.  I see that the log message was changed in ed866afc8c8bc9ce8be5f7ffb17c696b11980397 and b1a4f1bf7f0e214a900f9c3e87bc49212d0f95bb and is now "Error while traversing In-progress folder for the tenant".  I'm going to mark the conversation as resolved and open a fresh conversation on it.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchNameProvider.java:
```diff
@@ -71,4 +82,17 @@ public static boolean validateFileName(final String fileName)
+        final String noThreadIdFileName = noCreationTimeFileName.substring(endOfThreadId + 1);
+        final int endOfServiceId = noThreadIdFileName.indexOf('-');
+        String serviceID = noThreadIdFileName.substring(0,endOfServiceId);
+        return threadID+"-"+serviceID;
```
This needs a new class for returning it rather than munging it together into a string.  That will only force you to have to pull it apart again later on.

dermot-hardy replied:
I see you attempted to fix this with `ExtractedThreadAndServiceIDs`.  Unfortunately you undermined the effort by introducing and using the `getCombinedIDs()` method.

I've resolved this properly in this commit:
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
         return getPathForTenant(tenantId).resolve(INPROGRESS_FOLDER);
     }
 
+    public void monitorBatchProgressInFileSystem(final TenantId tenantId, final BatchId batchId, String threadID)
```
All parameters should be `final`:

```suggestion
    public void monitorBatchProgressInFileSystem(final TenantId tenantId, final BatchId batchId, final String threadID)
```

raghavgopal replied:
Done.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+    {
+        final Path pathForBatches = getTenantInprogressDirectory(tenantId);
+        final long[] size = new long[1];
+        try (Stream<Path> fileStream = Files.list(pathForBatches)) {
```
I'm confused that we are streaming through the list again, re-looking for the folder.  Didn't we already stream through the list in the `getListOfInProgressThreadFromFileSystem()` function?  If we needed the full path then why did we throw it away only to have to re-find it again here?

raghavgopal replied:
Here we are only just going to track the progress of Batches being uploaded by a staging-service other than the current staging-service that is handling the getStatus API request.
Probably I need to rename the generic name `monitorBatchProgressInFileSystem` to something relevant. Can you coin any meaningful method name for this ?

dermot-hardy replied:
What was done here with the introduction of the `inProgressBatchesOfDifferentServices` variable was very damaging to the code.  It was much worse than when I originally made the comment about it streaming through the list again, and it was still iterating the list multiple times.

I've resolved this property in this commit:
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

The `getProgressReportFromDirectory()` method, which is what I renamed this method to, now takes the `Path` object, so that it doesn't need to re-find the relevant directory.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+                            !batch.getFileName().toString().contains(ServiceIdentifier.getServiceId()))
+                    .map(batch -> batch.resolve("files"))
+                    .forEach(batch -> {
+                        try (Stream<Path> filePath = Files.walk(batch)) {
```
Why `.walk()` rather than `.list()` here?  Can this contain folders?

raghavgopal replied:
Hmm.. I don't think there will be any sub-directories inside /files folder, I just included for a safe handling. I will put back to `list()`

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+    public void monitorBatchProgressInFileSystem(final TenantId tenantId, final BatchId batchId, String threadID)
+    {
+        final Path pathForBatches = getTenantInprogressDirectory(tenantId);
+        final long[] size = new long[1];
```
What's the point of the array, and why is it at this scope rather than inside the block where it is used?

raghavgopal replied:
As variable used in lambda expression should be final or effectively final, so to make it effectively final I had to use final one-element array.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+                    .forEach(batch -> {
+                        try (Stream<Path> filePath = Files.walk(batch)) {
+                            size[0] = filePath.mapToLong(file -> file.toFile().length()).sum();
+                        } catch (IOException e) {
+                            LOGGER.info("Folder might have been deleted or moved.");
```
What is this circumstance?  Is it that the batch completes in the middle of the traversal?  Is this the correct way to handle that?

raghavgopal replied:
Right, if the batch is completed during parsing, we will get IOException that file not found. I am now logging this in error log level.

dermot-hardy replied:
This was being caught in various places so I adjusted the code in commit 590f331c0db34ae520d5a2d8a635c62a07f8bd09 to let it bubble up so that it can be handled in the main `getStatus()` method.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+                                    trackerMap.put(threadID, tracker);
+                                }
+                            }
+                        } catch (IllegalStateException | IOException e) {
```
What are each of these two exception cases?

raghavgopal replied:
Just IOException is sufficient, handling accordingly.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+                                Instant prevTime = getLastModified(file);
+                                threadSleep();
+                                Instant currTime = getLastModified(file);
+                                if (ChronoUnit.MILLIS.between(prevTime, currTime) > 0) {
```
Is a tracker not required if there has been no modification of the file?

raghavgopal replied:
I have added another attribute in the tracker `isProgressOnTrack` to denote if there is no modification of the file.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +289,49 @@ public void cleanUpStaleInprogressBatches()
+            InProgressMetrics inProgressMetrics= new InProgressMetrics();
+            if (!batchThread.contains(ServiceIdentifier.getServiceId()))
+                batchPathProvider.monitorBatchProgressInFileSystem(tenantId, batchId, batchThread);
+            if(trackerMap.get(batchThread) != null) {
+                inProgressMetrics.setLastModifiedDate(trackerMap.get(batchThread).getLastModifiedTime());
+                inProgressMetrics.setBytesReceived(trackerMap.get(batchThread).getNumberOfBytesReceived());
+                inProgressMetrics.setBytesPerSecond(trackerMap.get(batchThread).getFileUploadRate());
```
Why are we repeatedly calling `trackerMap.get(batchThread)` on these lines?  Can't this just be done once?

raghavgopal replied:
Yes, corrected.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/utils/Tracker.java:
```diff
@@ -0,0 +1,77 @@
+        this.numberOfBytesReceived = numberOfBytesReceived;
+    }
+
+    public Long getFileUploadRate() {
```
I know the unit is bytes/second but there is no indication of that in the method name or in any Javadoc.

raghavgopal replied:
Method name is renamed accordingly.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +289,49 @@ public void cleanUpStaleInprogressBatches()
+        // Remove the entry from TrackerMap for batch processed by different staging service
+        trackerMap.entrySet().removeIf(entry ->!entry.getKey().contains(ServiceIdentifier.getServiceId()));
+        // Build and return the Response
+        inProgress.setNumberOfBatches(inProgressMetricsList.size());
```
I think we should drop this number of batches from the response payload.  The caller can easily see the size of the array themselves so it's really not adding anything.

raghavgopal replied:
Done.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -109,6 +126,68 @@ public Path getTenantInprogressDirectory(final TenantId tenantId)
+    {
+        try {
+            return Files.readAttributes(p, BasicFileAttributes.class).lastModifiedTime().toInstant();
+        } catch (IOException ioe) {
+            throw new IllegalStateException(ioe);
```
Why are we re-throwing this as an illegal exception?

raghavgopal replied:
Changed to handle as IOException itself in the called places.

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -138,4 +217,13 @@ public UnexpectedInvalidBatchIdException(final BatchId batchId)
+    {
+        try {
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
```
We shouldn't be catching this.  It should be allowed to propagate all the way up to `StagingController`.  At that point it should probably be mapped to a `503 Service Unavailable` response.

raghavgopal replied:
Done.

dermot-hardy replied:
You've caught it early and re-thrown it as a `ServiceUnavailableException`, which you've then re-caught in `StagingController`.  Just propagate the `InterruptedException` please.

raghavgopal replied:
Propagated to the controller.

raghavendran-gopalakrishnan committed:
Addressing review comments and code improvements

raghavendran-gopalakrishnan committed:
More code improvements

raghavendran-gopalakrishnan committed:
Update FileSystemDaoTest.java

raghavendran-gopalakrishnan committed:
Additional review comments

dermot-hardy committed:
Code review: Whitespace

dermot-hardy committed:
Code review: Import order

dermot-hardy committed:
Code review: Add missing final keywords

dermot-hardy commented on staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java:
```diff
@@ -18,20 +18,38 @@
+    public static final String MOVED_TO_PROGRESS = "The in-progress folder might have been moved to Completed.";
     private final Path basePath;
 
+    private final List<Path> inProgressBatchesOfDifferentServices = new ArrayList<>();
```
This has been introduced since the last code review and its introduction is a disaster.  It has severely damaged the code base by attempting to operate through a side-effect.  Furthermore you only ever add to it so it is both a memory leak and a performance issue as the number of paths build up.

dermot-hardy replied:
**Resolved in this commit:**
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

I scrapped this variable and updated the method (which is now called `getOtherServicesInProgressBatchDirectories()`) to directly return the list of paths which correspond to other services which are uploading the batch.  There is no need to also return the paths that correspond to uploads being done by this service since they can be looked up in the concurrent map object.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/FileSystemDao.java:
```diff
@@ -282,6 +290,45 @@ public void cleanUpStaleInprogressBatches()
+            if (!batchThread.contains(ServiceIdentifier.getServiceId())) {
+                batchProgressTracker.putAll(batchPathProvider.monitorBatchProgressOfDifferentServices(batchThread));
+            }
+            final Tracker tracker = batchProgressTracker.get(batchThread);
```
The code is effectively looking up by thread.  But how do you know that thread hasn't moved on to a different batch?  The thread could potentially have even moved on to dealing with a different tenant between the time you retrieved these ids and the time you are attempting to retrieve these progress reports.  So this is not guaranteed to be giving back progress reports for the correct batch.

dermot-hardy replied:
**Resolved in this commit:**
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

The map has been updated so that it is additionally keyed by tenant and batch rather than just by thread id.

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/utils/BatchProgressTracker.java:
```diff
@@ -0,0 +1,92 @@
+            final long diff = ChronoUnit.MILLIS.between(tracker.getUploadStartTime(), now);
+            if (diff > 0) {
+                final long speed = Math.floorDiv(totalBytesRead, diff);
+                tracker.setFileUploadRateInBytesPerSecond(speed * 1000);
+                tracker.setNumberOfBytesReceived(totalBytesRead);
+                tracker.setLastModifiedTime(now);
+                tracker.setProgressing(true);
```
Unless this is the first `update()` call, you are updating a `Tracker` object that is already in the `ConcurrentHashMap`.  But you are doing it piecemeal, one property at a time.  Another thread could be reading this information simultaneously, and therefore it could end up with an erroneous half-half report.  To do what you're trying to do you need to synchronize updates and reads in the `Tracker` object.  It would probably be better and easier to change the object to be immutable and to replace it in the map.

dermot-hardy replied:
**Resolved in this commit:**
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

The `Tracker` object has been updated to be immutable; all of its fields are now `final`.  The objects are now replaced in the map rather than being updated.

dermot-hardy committed:
Code review: Extensive refactoring

- Resolve various race conditions
- Move code out of staging-service-shared
- Simplify the code

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/statusreporting/BatchStatusProvider.java:
```diff
@@ -0,0 +1,134 @@
+                    list.add(batch);
+                }
+            }
+        } catch (final IOException e) {
+            LOGGER.error("Error while traversing In-progress folder for the tenant: " + tenantId, e);
+        }
```
I'm still concerned about this `IOException`.  It has now been changed from `info` to `error`.  The last time I asked about it the line was:

    LOGGER.info("The folder might have been moved or deleted.");

...but I want to go over it again.  When does this happen?  Have you ever actually seen it happening?  Can you make it happen on-demand?  Do you have the stack trace logged that I can look at?  It is really right that we simply log the error and return an empty list?

raghavgopal replied:
> @raghavendran-gopalakrishnan Unfortunately I can't afford another large review round on this so I've applied most of my own review comments: [0f8ce49...590f331](https://github.com/CAFDataProcessing/staging-service/compare/0f8ce49de8b26e66277fce949939a17c2b5f9844...590f331c0db34ae520d5a2d8a635c62a07f8bd09). However I haven't had time to test any of the changes that I've made so this still needs to go back to you for manual testing, and possibly to look at adding some more automated tests as well. Please test everything that I've adjusted because unfortunately I haven't run it at all.

Thank you for your very extensive review and refactoring! I went through the refactored changes and manual tests seems good. I am going to check for anymore automated tests required.

raghavgopal replied:
> I'm still concerned about this `IOException`. It has now been changed from `info` to `error`. The last time I asked about it the line was:
> 
> ```
> LOGGER.info("The folder might have been moved or deleted.");
> ```
> 
> ...but I want to go over it again. When does this happen? Have you ever actually seen it happening? Can you make it happen on-demand? Do you have the stack trace logged that I can look at? It is really right that we simply log the error and return an empty list?

The event of IOException happening here is not possible as it is very highly unlikely that the `in_progress` folder under the tenant folder gets deleted by any means. Perhaps that would be an error and needs to be propagated to the controller level and needs to throw an HTTP Error 500.

dermot-hardy replied:
If that is the situation, i.e. the `in_progress` folder itself being moved or deleted, then yes that is completely unexpected and should result in a 500.

dermot-hardy reviewed (commented):
@raghavendran-gopalakrishnan Unfortunately I can't afford another large review round on this so I've applied most of my own review comments: 0f8ce49de8b26e66277fce949939a17c2b5f9844...590f331c0db34ae520d5a2d8a635c62a07f8bd09.  However I haven't had time to test any of the changes that I've made so this still needs to go back to you for manual testing, and possibly to look at adding some more automated tests as well.  Please test everything that I've adjusted because unfortunately I haven't run it at all.

raghavendran-gopalakrishnan committed:
Handling IOException

raghavendran-gopalakrishnan committed:
Updated tests

dermot-hardy commented on staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/statusreporting/BatchStatusProvider.java:
```diff
@@ -0,0 +1,135 @@
+                }
+            }
+        } catch (final IOException e) {
+            LOGGER.error("Error while traversing In-progress folder for the tenant: " + tenantId, e);
```
This line should have been removed when the exception was added, because the exception is caught and logged.  Leaving this log statement will cause double-logging of the error.

(The logging in the `getStatus()` method above is because there may be multiple exceptions in the collection and we are only going to throw one of them.  There is double-logging of the first one but it is not worth the extra code to prevent it.)

dermot-hardy replied:
**Resolved in this commit:**  
[b60a18bec0e53e0b468bc8604ad947afbd22107e] Drop the double-logging

dermot-hardy committed:
Code review: Whitespace

dermot-hardy committed:
Drop the double-logging
